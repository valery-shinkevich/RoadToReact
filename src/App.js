import React, { Component } from "react"
import axios from 'axios'
import "./App.css"
import Search from "./Search"
import Table from "./Table"
import Button from "./Button"

const DEFAULT_QUERY = "redux"
const DEFAULT_HPP = "5"

const PATH_BASE = "https://hn.algolia.com/api/v1"
const PATH_SEARCH = "/search"
const PARAM_SEARCH = "query="
const PARAM_PAGE = "page="
const PARAM_HPP = "hitsPerPage="

class App extends Component {

  _isMounted = false

  state = {
      results: null,
      searchKey: "",
      searchTerm: DEFAULT_QUERY,
      hitsPerPage: DEFAULT_HPP,
      error: null,
    }

  render() {
    const { searchTerm, results, searchKey, hitsPerPage, error } = this.state
    const resultByKey = results && results[searchKey]
    const page = (resultByKey && resultByKey.page) || 0
    const list = (resultByKey && resultByKey.hits) || []

    return (
      <div className="page">
        <div className="interactions">
          <Search
            value={searchTerm}
            hitsPerPage={hitsPerPage}
            onChange={this.onSearchChange}
            onSubmit={this.onSearchSubmit}
            onHppChange={this.onHppChange}
          >
            Поиск
          </Search>
        </div>
        { error
          ? <div className="interactions"><p>Что-то пошло не так...</p></div>
          : <Table list={list} onDismiss={this.onDismiss} />}
            <div className="interactions">
              <Button
                onClick={() => this.fetchSearchTopStories(searchKey, hitsPerPage, page + 1)}
              >
              Больше историй
              </Button>
            </div>
      </div>
    )
  }

  componentDidMount() {
    this._isMounted = true

    const { searchTerm, hitsPerPage } = this.state
    this.setState({
      searchKey: searchTerm
    })
    this.fetchSearchTopStories(searchTerm, hitsPerPage)
  }

  componentWillUnmount(){
    this._isMounted = false
  }

  onDismiss = (id) => {
    const { searchKey, results } = this.state
    const { hits, page } = results && results[searchKey]
    const isNotId = item => item.objectID !== id
    const updatedHits = hits.filter(isNotId)
    this.updateStateResults(results, searchKey, updatedHits, page)
  }

  onSearchSubmit = (event) => {
    const {searchTerm,hitsPerPage} = this.state
    
    this.setState({searchKey: searchTerm})
    
    if (this.needsToSearchTopStories(searchTerm)) {
      this.fetchSearchTopStories(searchTerm, hitsPerPage)
    }
    
    event.preventDefault()
  }

  onSearchChange = (event) => {
    this.setState({
      searchTerm: event.target.value
    })
  }

  onHppChange = (event) => {
    this.setState({
      hitsPerPage: event.target.value
    })
  }

  needsToSearchTopStories = (searchTerm) => {
    return !this.state.results[searchTerm]
  }

  fetchSearchTopStories = (term, hitsPerPage, page = 0) => {
    axios.get(`${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${term}&${PARAM_PAGE}${page}&${PARAM_HPP}${hitsPerPage}`)
      .then(result => this._isMounted && this.setSearchTopStories(result.data))
      .catch(error => this._isMounted && this.setState({ error }))
  }

  setSearchTopStories = (founded) => {
    const { hits, page } = founded
    const { searchKey, results } = this.state
    const oldHits =
      (results && results[searchKey] && results[searchKey].hits) || []
    const newHits = [...oldHits, ...hits]

    this.updateStateResults(results, searchKey, newHits, page)
  }

  updateStateResults = (results, searchKey, updatedHits, page) => {
    this.setState({
      results: {
        ...results,
        [searchKey]: {
          hits: updatedHits,
          page
        }
      }
    })
  }
}

export default App
