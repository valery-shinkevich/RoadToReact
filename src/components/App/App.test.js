import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'

import App from './App'

describe('App', () => {
  
  it('renders without crashing', () => {
  const div = document.createElement('div')
  ReactDOM.render(<App />, div)
  ReactDOM.unmountComponentAtNode(div)
  })
  
  test('есть корректный снимок', () => {
    const component = renderer.create(
      <App />
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

})