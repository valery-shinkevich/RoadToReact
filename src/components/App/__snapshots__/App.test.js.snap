// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App есть корректный снимок 1`] = `
<div
  className="page"
>
  <div
    className="interactions"
  >
    <form
      onSubmit={[Function]}
    >
      <input
        onChange={[Function]}
        type="text"
        value="redux"
      />
      Строк на странице:
      <input
        onChange={[Function]}
        style={
          Object {
            "width": 20,
          }
        }
        type="text"
        value="5"
      />
      <button
        type="submit"
      >
        Поиск
      </button>
    </form>
  </div>
  <div
    className="table"
  />
  <div
    className="interactions"
  >
    <button
      className=""
      onClick={[Function]}
      type="button"
    >
      Больше историй
    </button>
  </div>
</div>
`;
